lib/stdlib/src/lists.erl

```diff
@@ -260,15 +260,19 @@ seq_loop(0, _, L) ->
       Seq :: [integer()].
 
 seq(First, Last, Inc) 
-    when is_integer(First), is_integer(Last), is_integer(Inc) -> 
+    when is_integer(First), is_integer(Last), is_integer(Inc),
+        (Inc > 0 andalso First - Inc =< Last) orelse
+        (Inc < 0 andalso First - Inc >= Last) orelse
+        (Inc =:= 0 andalso First =:= Last) -> 
     if
-        Inc > 0, First - Inc =< Last;
-        Inc < 0, First - Inc >= Last ->
+        Inc =/= 0 ->
             N = (Last - First + Inc) div Inc,
             seq_loop(N, Inc*(N-1)+First, Inc, []);
-        Inc =:= 0, First =:= Last ->
+        Inc =:= 0 ->
             seq_loop(1, First, Inc, [])
-    end.
+    end;
+seq(First, Last, Inc) ->
+    erlang:error(badarg, [First, Last, Inc], [{error_info, #{module => erl_stdlib_errors}}]).
```

Comment:
After checking the disassembly, it seems the compiler doesn't manage to optimize the `if` very well. Can you hoist it up to a function clause instead? e.g.

```erlang
seq(First, Last, Inc)
    when is_integer(First), is_integer(Last), is_integer(Inc),
        (Inc > 0 andalso First - Inc =< Last) orelse
        (Inc < 0 andalso First - Inc >= Last) ->
    N = (Last - First + Inc) div Inc,
    seq_loop(N, Inc * (N - 1) + First, Inc, []);
seq(Same, Same, 0) when is_integer(Same) ->
    [Same];
seq(First, Last, Inc) ->
    erlang:error(badarg, [First, Last, Inc],
                 [{error_info, #{module => erl_stdlib_errors}}]).
```